torchgen/executorch/api/cpp.py

```diff
@@ -0,0 +1,409 @@
+from typing import List, Optional, Sequence, Set, Union
+
+from torchgen import local
+from torchgen.api.types import (
+    ArgName,
+    ArrayCType,
+    BaseCType,
+    Binding,
+    ConstRefCType,
+    CType,
+    longT,
+    MutRefCType,
+    NamedCType,
+    SpecialArgName,
+    TupleCType,
+    VectorCType,
+    voidT,
+)
+from torchgen.model import (
+    Argument,
+    Arguments,
+    BaseTy,
+    BaseType,
+    FunctionSchema,
+    ListType,
+    NativeFunction,
+    OptionalType,
+    Return,
+    SelfArgument,
+    TensorOptionsArguments,
+    Type,
+)
+from torchgen.utils import assert_never
+
+from .types import (
+    ArrayRefCType,
+    BaseTypeToCppMapping,
+    intArrayRefT,
+    OptionalCType,
+    scalarT,
+    tensorListT,
+    tensorT,
+)
+
+
+# This file describes the translation of JIT schema to the public C++
+# API, which is what people use when they call functions like at::add,
+# as well as native function API, which is the signature of kernels.
+# Notice that in Executorch CppSignature is the same as NativeSignature.
+#
+# Prominent characteristics of the C++ API:
+#
+#   - dtype, layout, device and pin_memory are collected into
+#     a single C++ type TensorOptions  (the native functions API
+#     also has this, but tensor options is really most relevant
+#     for the C++ API; it makes calling kwarg factory functions
+#     pleasant). Executorch doesn't have TensorOptions support,
+#     we keep the logic here to be compatible with torchgen.api.
+#     types.CppSignature, so that we can do stuff like ATen mode.
+#
+#   - defaulting lives here (in fact, the dispatcher is completely
+#     oblivious of defaults!)
```

Comment:
Instead of copy pasting the old comment, please talk about how this differs from the regular cpp api